File: Algorithms/0554_Brick_Wall/Java/BrickWallSolution1.java

// Space: O(n)
// Time: O(n)

public class Solution {
    public int leastBricks(List<List<Integer>> wall) {
        int height = wall.size();
        if (height == 1) {
            return wall.get(0).size() == 1 ? 1 : 0;
        }

        List<List<Integer>> cache = new ArrayList<>();
        Map<Integer, Integer> count = new HashMap<>();
        int maxEdge = -1, res = -1, finalRes = 0;

        for (List<Integer> row : wall) {
            List<Integer> temp = new ArrayList<>();
            for (int i = 1; i < row.size(); i++) {
                int edge = row.subList(0, i).stream().mapToInt(Integer::intValue).sum();
                temp.add(edge);
            }
            cache.add(temp);
        }
        for (List<Integer> row : cache) {
            for (int e : row) {
                if (count.containsKey(e)) {
                    count.put(e, count.get(e) + 1);
                } else {
                    count.put(e, 1);
                }
                if (count.get(e) > maxEdge) {
                    maxEdge = count.get(e);
                    res = e;
                }
            }
        }
        if (res == -1) return height;
        for (List<Integer> row : cache) {
            if (!row.contains(res)) finalRes += 1;
        }
        return finalRes;
    }
}

File: Algorithms/0554_Brick_Wall/Python/Brick_Wall_Solution_1.py

# Space: O(n)
# Time: O(n)

import collections


class Solution:
    def leastBricks(self, wall):
        height = len(wall)
        if height == 1:
            return 0 if len(wall[0]) > 1 else 1
        cache = []
        count = collections.defaultdict(int)
        res, final_res = -1, 0

        for row in wall:
            edge = []
            for i in range(1, len(row)):
                edge.append(sum(row[:i]))
            cache.append(edge)

        for row in cache:
            for e in row:
                count[e] += 1
                if res == -1:
                    res = e
                else:
                    res = e if count[e] > count[res] else res

        if res == -1: return height

        for row in cache:
            if res not in row:
                final_res += 1

        return final_res




